crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -300,10 +283,96 @@ impl Persistence {
             prices,
-            deadline,
             surplus_capturing_jit_order_owners,
         })
     }
+
+    /// Get competition winner.
+    pub async fn get_competition_winner(
+        &self,
+        auction_id: domain::auction::Id,
+    ) -> Result<domain::competition::Solution, error::Winner> {
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["get_competition_winner"])
+            .start_timer();
+
+        let mut ex = self
+            .postgres
+            .pool
+            .begin()
+            .await
+            .map_err(error::Winner::DatabaseError)?;
+
+        let competition = database::settlement_scores::fetch(&mut ex, auction_id)
+            .await
+            .map_err(error::Winner::DatabaseError)?
+            .ok_or(error::Winner::Missing)?;
+
+        let winner = H160(competition.winner.0).into();
+        let score = competition::Score::new(
+            big_decimal_to_u256(&competition.winning_score)
+                .ok_or(error::Winner::InvalidScore(anyhow::anyhow!(
+                    "database score"
+                )))?
+                .into(),
+        )
+        .map_err(|_| error::Winner::InvalidScore(anyhow::anyhow!("zero score")))?;
+
+        let solution = {
+            // TODO: stabilize the solver competition table to get promised solution.
+            let solver_competition = database::solver_competition::load_by_id(&mut ex, auction_id)
+                .await
+                .map_err(error::Winner::DatabaseError)?
+                .ok_or(error::Winner::Missing)?;
+            let competition: model::solver_competition::SolverCompetitionDB =
+                serde_json::from_value(solver_competition.json)
+                    .context("deserialize SolverCompetitionDB")
+                    .map_err(error::Winner::SolverCompetition)?;
+            let winning_solution = competition.solutions.last().ok_or(error::Winner::Missing)?;
+            let mut orders = HashMap::new();
+            for order in winning_solution.orders.iter() {
+                match order {
+                    model::solver_competition::Order::Colocated {
+                        id,
+                        sell_amount,
+                        buy_amount,
+                    } => {
+                        orders.insert(
+                            domain::OrderUid(id.0),
+                            competition::TradedAmounts {
+                                sell: (*sell_amount).into(),
+                                buy: (*buy_amount).into(),
+                            },
+                        );
+                    }
+                    model::solver_competition::Order::Legacy {
+                        id: _,
+                        executed_amount: _,
+                    } => {
+                        return Err(error::Winner::SolverCompetition(anyhow::anyhow!(
+                            "Legacy order"
+                        )))
```

Comment:
This should be treated equivalent to NotFound in my opinion (we can document that this method only works for auctions that run in co-located mode)

Reply:
resolved